src/coreclr/jit/optimizer.cpp

```diff
@@ -9335,6 +9337,262 @@ bool OptBoolsDsc::optOptimizeBoolsCondBlock()
+//      *  ASG x,z
+//
+//
+//      Example 3:
+//          If ( a > b || c == d || e < f ) { x = y; }
+//      The first pass of the optimization will combine two of the conditions. The
+//      second pass will then combine remaining condition the earlier chain.
+//
+//      ------------ BB01 -> BB03 (cond), succs={BB03,BB04}
+//      *  JTRUE (NE (OR ((NE (OR (NE c,d), (GE e,f)), 0), (LE a,b))), 0)
+//
+//      ------------ BB03, preds={BB01} succs={BB04}
+//      *  ASG x,y
+//
+//
+//     This optimization means that every condition within the IF statement is always evaluated,
+//     as opposed to stopping at the first positive match.
+//     Theoretically there is no maximum limit on the size of the generated chain. Therefore cost
+//     checking is used to limit the maximum number of conditions that can be chained together.
+//
+bool OptBoolsDsc::optOptimizeCompareChainCondBlock()
+{
+    assert(m_b1 != nullptr && m_b2 != nullptr && m_b3 == nullptr);
+    m_t3 = nullptr;
+
+    bool foundEndOfOrConditions = false;
+    if (m_b1->bbNext == m_b2 && m_b1->bbJumpDest == m_b2->bbNext && m_b1->bbJumpDest->bbNext == m_b2->bbJumpDest)
+    {
+        // Found the end of two (or more) conditions being ORed together.
+        // The final condition has been inverted.
+        foundEndOfOrConditions = true;
+    }
+    else if (!(m_b1->bbNext == m_b2 && m_b1->bbJumpDest == m_b2->bbJumpDest))
+    {
+        return false;
+    }
+
+    Statement* const s1 = optOptimizeBoolsChkBlkCond();
+    if (s1 == nullptr)
+    {
+        return false;
+    }
+    Statement* s2 = m_b2->firstStmt();
+
+    assert(m_testInfo1.testTree->OperIs(GT_JTRUE));
+    GenTree* cond1 = m_testInfo1.testTree->gtGetOp1();
+    assert(m_testInfo2.testTree->OperIs(GT_JTRUE));
+    GenTree* cond2 = m_testInfo2.testTree->gtGetOp1();
+
+    // Ensure both conditions are suitable.
+    if (!cond1->OperIsCompare() || !cond2->OperIsCompare())
+    {
+        return false;
+    }
+
+    // Ensure there are no additional side effects.
+    if ((cond1->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0 ||
+        (cond2->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+    {
+        return false;
+    }
+
+    // Integer compares only for now (until support for Arm64 fccmp instruction is added)
+    if (varTypeIsFloating(cond1->gtGetOp1()->TypeGet()) || varTypeIsFloating(cond2->gtGetOp1()->TypeGet()))
+    {
+        return false;
+    }
+
+    // Check for previously optimized compare chains.
+    bool op1IsTestCond;
+    bool op2IsTestCond;
+    bool op1IsCondChain = FindCompareChain(cond1, &op1IsTestCond);
+    bool op2IsCondChain = FindCompareChain(cond2, &op2IsTestCond);
+
+    // Specifically for Arm64, avoid cases where optimizations in lowering will produce better
+    // code than optimizing here. Specificially:
+    // * CMP(AND(...), 0) will be turned into a TEST_ opcode.
+    // * Compares against zero will be optimized with cbz.
```

Comment:
I've been updating the code to also check for EQ/NE(AND(...,n),n) where n is a power of 2, which will optimize to tbz/tbnz.

I have a slight concern that this becomes a little awkward to manage, as newer tricks in lowering also need reflecting here. Don't think there is a better way around it - I don't think a common function in lowering would work. Maybe it just needs a comment in lowering.

Reply:
I'm not sure how viable this is, and I don't want to implement for this pull, but...

Ideally, there should be no cbz/tbz checks or other lowering checks in the optimize bools pass. Instead, lowering should be able to spot where two `CSET` conditions are combined together then used in a jump, and replace it with two jumps

Simplified example:
eg:
```
CMP
CSET
CMP
CSET
AND
JCC LABEL
```

gets lowered to:

```
CMP
JCC LABEL
CMP
JCC LABEL
```

I'm not sure how tricky it'd be to ensure it captures all the tbz/tbnz and cbz/cbnz cases.

Reply:
It would probably produce smaller code, but I would be concerned that it runs into the opposite problem about knowing whether or not this is an optimization. If the user deliberately writes code like `((x & 4) == 0) & ((x & 0x2) != 0)` then they probably expect branchless codegen and that the first condition is nearly always true. Is it then a good idea to transform it into a branch?

Neither GCC nor clang emit extra branches, though they seem to employ some other tricks to make the codegen more efficient and avoid the `tst + cset`: https://godbolt.org/z/4MxhW1114